src/pynwb/core.py

```diff
@@ -318,6 +319,8 @@ def __getitem__(self, args):
         return self.data[args]
 
     def append(self, arg):
+        if isinstance(self.data, HDMFDataset) or isinstance(self.data, Dataset):
+            self.__data = self.data[()]
```

Comment:
You could avoid reading the dataset into memory by reshaping the dataset, and then adding the new data.